src/control/lib/hardware/hwloc/provider.go

```diff
@@ -15,6 +15,43 @@ import (
 	"github.com/daos-stack/daos/src/control/logging"
 )
 
+type ctxKey string
+
+const (
+	topoKey    ctxKey = "hwlocTopology"
+	cleanupKey ctxKey = "hwlocFreeTopology"
+)
+
+// CacheContext adds a cache of the hwloc topology to the provided context.
+func CacheContext(parent context.Context, log logging.Logger) (context.Context, error) {
+	topo, cleanup, err := NewProvider(log).initTopology()
+	if err != nil {
+		return parent, err
```

Comment:
nit: Would be better to `return nil, err`, as callers should be checking the error and not expecting a valid context in the case of an error.